pubsub/kafka/kafka.go

```diff
@@ -24,32 +26,50 @@ import (
 
 type PubSub struct {
 	kafka  *kafka.Kafka
-	topics map[string]bool
+	topics map[string]pubsub.Handler
+	lock   sync.RWMutex
 }
 
 func (p *PubSub) Init(metadata pubsub.Metadata) error {
-	p.topics = make(map[string]bool)
 	return p.kafka.Init(metadata.Properties)
 }
 
 func (p *PubSub) Subscribe(req pubsub.SubscribeRequest, handler pubsub.Handler) error {
-	topics := p.addTopic(req.Topic)
+	// Add the topic to the list of those we subscribe to
+	p.lock.Lock()
+	p.topics[req.Topic] = handler
+	topics := p.topicList()
+	p.lock.Unlock()
 
-	return p.kafka.Subscribe(topics, req.Metadata, newSubscribeAdapter(handler).adapter)
+	return p.kafka.Subscribe(context.Background(), topics, req.Metadata, p.handleMessage)
 }
 
-func (p *PubSub) addTopic(newTopic string) []string {
-	// Add topic to our map of topics
-	p.topics[newTopic] = true
+func (p *PubSub) handleMessage(ctx context.Context, event *kafka.NewEvent) error {
+	// Get the handler
+	p.lock.RLock()
```

Comment:
Is this read lock needed if `Subscribe` is called for all topics before `handleMessage` is ever executed?

Reply:
Actually in this case, no. But I backported this change from the work I've been doing on #1756 and in that case it's probably necessary if the subscription is being removed.

I can delete it from here for now (I have to make a few more changes to this PR per Sky's feedback anyways)